File: update_csrt/test_basketball.py
"""
Quick test on Basketball sequence
Outputs AUC and Precision@20px
"""

import sys
import cv2
import numpy as np
from pathlib import Path
import torch

sys.path.append(str(Path(__file__).parent))

from tracker import PCSRTTracker
from config import PCSRTConfig

def compute_iou(box1, box2):
    """Compute IoU between two boxes [x, y, w, h]"""
    x1, y1, w1, h1 = box1
    x2, y2, w2, h2 = box2
    
    xi1 = max(x1, x2)
    yi1 = max(y1, y2)
    xi2 = min(x1+w1, x2+w2)
    yi2 = min(y1+h1, y2+h2)
    
    if xi2 <= xi1 or yi2 <= yi1:
        return 0.0
    
    inter = (xi2 - xi1) * (yi2 - yi1)
    union = w1*h1 + w2*h2 - inter
    return inter / union

def compute_center_distance(box1, box2):
    """Compute center distance"""
    c1x, c1y = box1[0] + box1[2]/2, box1[1] + box1[3]/2
    c2x, c2y = box2[0] + box2[2]/2, box2[1] + box2[3]/2
    return np.sqrt((c1x-c2x)**2 + (c1y-c2y)**2)

def load_groundtruth(gt_file):
    """Load ground truth annotations"""
    gt = []
    with open(gt_file, 'r') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            # Try comma or space or tab separated
            if ',' in line:
                parts = [float(x) for x in line.replace('\t', ',').split(',') if x.strip()]
            else:
                parts = [float(x) for x in line.split()]
            if len(parts) >= 4:
                # Handle format: x,y,w,h
                gt.append([parts[0], parts[1], parts[2], parts[3]])
    return np.array(gt)

def main():
    # Basketball sequence path
    seq_path = Path("../otb100/OTB-dataset/OTB100/Basketball/Basketball")
    
    if not seq_path.exists():
        print(f"Error: Basketball sequence not found at {seq_path}")
        print(f"Absolute: {seq_path.absolute()}")
        return
    
    print("="*80)
    print("Testing Updated CSRT Tracker on Basketball")
    print("="*80)
    
    # Load ground truth
    gt_file = seq_path / "groundtruth_rect.txt"
    if not gt_file.exists():
        print(f"Error: Ground truth not found at {gt_file}")
        return
    
    gt_boxes = load_groundtruth(gt_file)
    print(f"\nLoaded {len(gt_boxes)} ground truth boxes")
    
    # Load images
    img_dir = seq_path / "img"
    if not img_dir.exists():
        print(f"Error: Image directory not found at {img_dir}")
        return
    
    img_files = sorted(img_dir.glob("*.jpg"))
    if not img_files:
        img_files = sorted(img_dir.glob("*.png"))
    
    print(f"Found {len(img_files)} images")
    
    num_frames = min(len(gt_boxes), len(img_files))
    print(f"Will process {num_frames} frames\n")
    
    # Initialize tracker
    print("Initializing tracker...")
    config = PCSRTConfig()
    
    # Use CPU
    device = torch.device('cpu')
    tracker = PCSRTTracker(config).to(device).eval()
    
    # First frame
    frame = cv2.imread(str(img_files[0]))
    if frame is None:
        print(f"Error: Cannot read first frame: {img_files[0]}")
        return
    
    print(f"Frame size: {frame.shape}")
    print(f"Initial bbox: {gt_boxes[0]}")
    
    # Initialize with ground truth
    init_bbox = gt_boxes[0]
    tracker.initialize(frame, init_bbox)
    
    print("\nTracking...")
    
    # Track remaining frames
    overlaps = []
    distances = []
    
    for i in range(1, num_frames):
        frame = cv2.imread(str(img_files[i]))
        if frame is None:
            print(f"Warning: Cannot read frame {i}")
            break
        
        # Track
        pred_bbox, confidence = tracker.track(frame)
        gt_bbox = gt_boxes[i]
        
        # Compute metrics
        iou = compute_iou(pred_bbox, gt_bbox)
        dist = compute_center_distance(pred_bbox, gt_bbox)
        
        overlaps.append(iou)
        distances.append(dist)
        
        # Progress
        if (i+1) % 50 == 0 or i == num_frames - 1:
            print(f"  Frame {i+1}/{num_frames} - IoU: {iou:.3f}, Dist: {dist:.1f}px")
    
    # Compute final metrics
    overlaps = np.array(overlaps)
    distances = np.array(distances)
    
    print("\n" + "="*80)
    print("Results")
    print("="*80)
    
    # AUC (Area Under Curve for success plot)
    thresholds = np.linspace(0, 1, 50)
    success_rates = [(overlaps > t).mean() for t in thresholds]
    auc = np.mean(success_rates)
    
    # Precision @ 20px
    precision_20 = (distances <= 20).mean()
    
    # Additional metrics
    mean_iou = overlaps.mean()
    mean_dist = distances.mean()
    
    print(f"Frames tracked:  {len(overlaps)}")
    print(f"AUC:             {auc:.3f}")
    print(f"Precision@20px:  {precision_20:.3f}")
    print(f"Mean IoU:        {mean_iou:.3f}")
    print(f"Mean Distance:   {mean_dist:.1f}px")
    print("="*80)

if __name__ == "__main__":
    main()

File: pure_csrt_detail/src/test_basketball.cpp
#include <opencv2/opencv.hpp>

#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "../inc/csrt_tracker.hpp"

namespace {

std::vector<cv::Rect> LoadGroundTruth(const std::string &filename) {
    std::vector<cv::Rect> gt;
    std::ifstream file(filename);
    std::string line;
    while (std::getline(file, line)) {
        std::replace(line.begin(), line.end(), ',', ' ');
        std::replace(line.begin(), line.end(), '\t', ' ');
        std::istringstream iss(line);
        int x = 0;
        int y = 0;
        int w = 0;
        int h = 0;
        if (iss >> x >> y >> w >> h) {
            gt.emplace_back(x, y, w, h);
        }
    }
    return gt;
}

float ComputeIoU(const cv::Rect &a, const cv::Rect &b) {
    int x1 = std::max(a.x, b.x);
    int y1 = std::max(a.y, b.y);
    int x2 = std::min(a.x + a.width, b.x + b.width);
    int y2 = std::min(a.y + a.height, b.y + b.height);
    if (x2 <= x1 || y2 <= y1) {
        return 0.0f;
    }
    float inter = static_cast<float>((x2 - x1) * (y2 - y1));
    float union_area = static_cast<float>(a.width * a.height + b.width * b.height) - inter;
    return inter / union_area;
}

float ComputeCenterDist(const cv::Rect &a, const cv::Rect &b) {
    cv::Point2f ca(a.x + a.width / 2.0f, a.y + a.height / 2.0f);
    cv::Point2f cb(b.x + b.width / 2.0f, b.y + b.height / 2.0f);
    return static_cast<float>(cv::norm(ca - cb));
}

}  // namespace

int main() {
    std::string base_path = "E:/SourceCode/C2P/Project/CSRTResearch/otb100/OTB-dataset/OTB100/Basketball/Basketball";

    std::cout << "================================================================================\n";
    std::cout << "Testing CSRT From Scratch on Basketball (First 50 frames)\n";
    std::cout << "================================================================================\n";

    auto gt_boxes = LoadGroundTruth(base_path + "/groundtruth_rect.txt");
    std::cout << "\nLoaded " << gt_boxes.size() << " GT boxes\n";

    std::vector<std::string> img_files;
    for (int i = 1; i <= 50 && i <= static_cast<int>(gt_boxes.size()); ++i) {
        char filename[256];
        std::snprintf(filename, sizeof(filename), "%s/img/%04d.jpg", base_path.c_str(), i);
        img_files.push_back(filename);
    }

    csrt::CsrtTracker tracker;
    cv::Mat frame = cv::imread(img_files[0]);
    if (frame.empty()) {
        std::cerr << "Cannot read first frame!\n";
        return 1;
    }

    std::cout << "Frame size: " << frame.size() << "\n";
    std::cout << "Init bbox: " << gt_boxes[0] << "\n";

    if (!tracker.Init(frame, gt_boxes[0])) {
        std::cerr << "Initialization failed!\n";
        return 1;
    }

    std::cout << "\nTracking...\n";

    std::vector<float> ious;
    std::vector<float> dists;

    for (size_t i = 1; i < std::min(img_files.size(), gt_boxes.size()); ++i) {
        frame = cv::imread(img_files[i]);
        if (frame.empty()) {
            break;
        }

        cv::Rect pred_box;
        if (!tracker.Update(frame, pred_box)) {
            std::cout << "  Frame " << (i + 1) << " - Tracking failed!\n";
            continue;
        }

        float iou = ComputeIoU(pred_box, gt_boxes[i]);
        float dist = ComputeCenterDist(pred_box, gt_boxes[i]);
        ious.push_back(iou);
        dists.push_back(dist);

        if ((i + 1) % 10 == 0 || i == img_files.size() - 1) {
            std::cout << "  Frame " << (i + 1) << "/" << img_files.size()
                      << " - IoU: " << std::fixed << std::setprecision(3) << iou
                      << ", Dist: " << std::fixed << std::setprecision(1) << dist << "px\n";
        }
    }

    float mean_iou = 0.0f;
    float mean_dist = 0.0f;
    int precision_20 = 0;

    for (size_t i = 0; i < ious.size(); ++i) {
        mean_iou += ious[i];
        mean_dist += dists[i];
        if (dists[i] <= 20.0f) {
            precision_20++;
        }
    }

    mean_iou /= ious.size();
    mean_dist /= dists.size();
    float p20 = precision_20 / static_cast<float>(ious.size());

    int num_thresholds = 50;
    float auc = 0.0f;
    for (int t = 0; t <= num_thresholds; ++t) {
        float threshold = t / static_cast<float>(num_thresholds);
        int count = 0;
        for (float iou : ious) {
            if (iou > threshold) {
                count++;
            }
        }
        auc += count / static_cast<float>(ious.size());
    }
    auc /= (num_thresholds + 1);

    std::cout << "\n================================================================================\n";
    std::cout << "Results (50 frames)\n";
    std::cout << "================================================================================\n";
    std::cout << "AUC:             " << std::fixed << std::setprecision(3) << auc << "\n";
    std::cout << "Precision@20px:  " << std::fixed << std::setprecision(3) << p20 << "\n";
    std::cout << "Mean IoU:        " << std::fixed << std::setprecision(3) << mean_iou << "\n";
    std::cout << "Mean Distance:   " << std::fixed << std::setprecision(1) << mean_dist << "px\n";
    std::cout << "================================================================================\n";

    return 0;
}
